File: composeApp/src/commonMain/kotlin/dev/mks/duskread/ui/theme/Theme.kt
package dev.mks.duskread.ui.theme

import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.darkColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color

// "Paper Black" — a page, not a screen; ink, not a glow.
private val DarkScheme = schemeOf(DesignTokens.paperBlack)

// "Ink" — the same page, with the ink drained out of it.
private val MonoScheme = schemeOf(DesignTokens.ink)

/**
 * Builds a Material scheme from a [Palette].
 */
private fun schemeOf(palette: Palette) = darkColorScheme(
    primary = Color(palette.primary),
    onPrimary = Color(palette.onPrimary),
    primaryContainer = Color(palette.primaryContainer),
    onPrimaryContainer = Color(palette.onPrimaryContainer),
    background = Color(palette.background),
    onBackground = Color(palette.onBackground),
    surface = Color(palette.surface),
    onSurface = Color(palette.onSurface),
    surfaceVariant = Color(palette.surfaceVariant),
    onSurfaceVariant = Color(palette.onSurfaceVariant),
    surfaceContainer = Color(palette.surfaceContainer),
    surfaceContainerHigh = Color(palette.surfaceContainerHigh),
    outline = Color(palette.outline),
    outlineVariant = Color(palette.outlineVariant),
    error = Color(palette.error),
)

/**
 * Both themes are dark; [mono] picks which.
 */
@Composable
fun DuskReadTheme(
    mono: Boolean = true,
    content: @Composable () -> Unit,
) {
    MaterialTheme(
        colorScheme = if (mono) MonoScheme else DarkScheme,
        typography = AlgoTypography(DuskReadFontFamily()),
        content = content,
    )
}

File: iosApp/iosApp/Design/Theme.swift
import ComposeApp
import SwiftUI

/// The two schemes, and the axis they differ on.
///
/// Not `@Environment(\.colorScheme)`: both schemes are dark, and what the
/// toggle changes is whether there is any hue at all. Reading the system's
/// light/dark setting would answer a question this app does not ask — and Ink,
/// the colourless one, is the default.
struct DuskTheme {
    let mono: Bool

    var palette: Palette { DesignTokens.shared.palette(mono: mono) }

    var primary: Color { Color(argb: palette.primary) }
    var onPrimary: Color { Color(argb: palette.onPrimary) }
    var primaryContainer: Color { Color(argb: palette.primaryContainer) }
    var onPrimaryContainer: Color { Color(argb: palette.onPrimaryContainer) }
    var background: Color { Color(argb: palette.background) }
    var onBackground: Color { Color(argb: palette.onBackground) }
    var surface: Color { Color(argb: palette.surface) }
    var onSurface: Color { Color(argb: palette.onSurface) }
    var surfaceVariant: Color { Color(argb: palette.surfaceVariant) }
    var onSurfaceVariant: Color { Color(argb: palette.onSurfaceVariant) }
    var surfaceContainer: Color { Color(argb: palette.surfaceContainer) }
    var surfaceContainerHigh: Color { Color(argb: palette.surfaceContainerHigh) }
    var outline: Color { Color(argb: palette.outline) }
    var outlineVariant: Color { Color(argb: palette.outlineVariant) }
    var error: Color { Color(argb: palette.error) }
}

private struct DuskThemeKey: EnvironmentKey {
    static let defaultValue = DuskTheme(mono: true)
}

extension EnvironmentValues {
    var dusk: DuskTheme {
        get { self[DuskThemeKey.self] }
        set { self[DuskThemeKey.self] = newValue }
    }
}

extension Color {
    /// Packed ARGB from Kotlin, which is how `DesignTokens` carries a colour —
    /// `Color` has no Obj-C representation, and a hex string would only be a
    /// number pretending not to be one.
    init(argb: Int64) {
        let value = UInt64(bitPattern: Int64(argb))
        self.init(
            .sRGB,
            red: Double((value >> 16) & 0xFF) / 255,
            green: Double((value >> 8) & 0xFF) / 255,
            blue: Double(value & 0xFF) / 255,
            opacity: Double((value >> 24) & 0xFF) / 255
        )
    }
}
